### ISSUE TITLE
Dockerized Promtheus run into "rename data/chunks_head/000021.tmp: Operation not permitted" error 

### ISSUE THREAD

**Issue by nico-1512** [prometheus/prometheus#13555]
### What did you do?

I'm creating a monitoring infrastructure hosting Grafana and Prometheus in a AWS EC2 machine. 
I've mounted an s3 bucket into my EC2 instance (with s3fs) and everythink seems to work just fine: I'm able to create, remove and update files and folders in the mounted bucket. To optimize costs and space I save Prometheus stuff (all the storaging) into my bucket. 

Here's my docker-compose.yml: 
```yml
version: '3.9'

services:
  
  grafana:
    image: grafana/grafana
    container_name: grafana
    restart: unless-stopped
    extra_hosts:
    - "host.docker.internal:host-gateway"
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /etc/timezone:/etc/timezone:ro        
      - ./grafana/data:/var/lib/grafana
    user: "0"                                             
    ports:
      - "3000:3000"

  prometheus:
    image: prom/prometheus:latest
    container_name: prometheus
    restart: unless-stopped
    network_mode: host
    volumes:
      - /home/ec2-user/s3-bucket/prometheus:/prometheus/
#      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#      Here is where I mount the bucket into my container
      - ./prometheus/:/etc/prometheus/
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
    privileged: true
    user: "1000:1000"
```

When I first create the container it just works fine until you restart manually it (e.g. I've updated the prometheus.yml file to scrape other metrics so to apply changes I've to recreate the container).  

I've also tried: 
`sudo chown -R 1000:1000 /home/ec2-user/s3-bucket `
`sudo chmod -R a+rwx  /home/ec2-user/s3-bucket `

### What did you expect to see?

I expect Prometheus container to read and write data in the bucket (since thanks to s3fs for the system the bucket is just another file system mounted partition). 

### What did you see instead? Under which circumstances?

I'm getting the following error (coming from `docker logs prometheus`): 

```
ts=2024-02-07T13:56:20.714Z caller=main.go:544 level=info msg="No time or size retention was set so using the default time retention" duration=15d
ts=2024-02-07T13:56:20.714Z caller=main.go:588 level=info msg="Starting Prometheus Server" mode=server version="(version=2.49.1, branch=HEAD, revision=43e14844a33b65e2a396e3944272af8b3a494071)"
ts=2024-02-07T13:56:20.714Z caller=main.go:593 level=info build_context="(go=go1.21.6, platform=linux/amd64, user=root@6d5f4c649d25, date=20240115-16:58:43, tags=netgo,builtinassets,stringlabels)"
ts=2024-02-07T13:56:20.714Z caller=main.go:594 level=info host_details="(Linux 6.1.72-96.166.amzn2023.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 17 00:42:52 UTC 2024 x86_64 ip-172-31-6-235.eu-west-1.compute.internal (none))"
ts=2024-02-07T13:56:20.714Z caller=main.go:595 level=info fd_limits="(soft=65536, hard=65536)"
ts=2024-02-07T13:56:20.714Z caller=main.go:596 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2024-02-07T13:56:20.801Z caller=web.go:565 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
ts=2024-02-07T13:56:20.802Z caller=main.go:1039 level=info msg="Starting TSDB ..."
ts=2024-02-07T13:56:20.805Z caller=tls_config.go:274 level=info component=web msg="Listening on" address=[::]:9090
ts=2024-02-07T13:56:20.805Z caller=tls_config.go:277 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
ts=2024-02-07T13:56:21.069Z caller=dir_locker.go:77 level=warn component=tsdb msg="A lockfile from a previous execution already existed. It was replaced" file=/prometheus/data/lock
ts=2024-02-07T13:56:21.624Z caller=head.go:606 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2024-02-07T13:56:21.624Z caller=head.go:687 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=23.625µs
ts=2024-02-07T13:56:21.624Z caller=head.go:695 level=info component=tsdb msg="Replaying WAL, this may take a while"
panic: 2 errors: replace file: rename data/chunks_head/000027.tmp data/chunks_head/000027: operation not permitted; unlinkat data/chunks_head/000027.tmp: operation not permitted

goroutine 425 [running]:
github.com/prometheus/prometheus/tsdb.handleChunkWriteError({0x3f3f3a0?, 0xc00097d038?})
	/app/tsdb/head_append.go:1432 +0x6e
github.com/prometheus/prometheus/tsdb/chunks.(*ChunkDiskMapper).WriteChunk(0xc0007b6000, 0xc000caad70?, 0x2760318?, 0xc001f4db90?, {0x3f73af8, 0xc00098e4e0}, 0x0?, 0x39d0540)
	/app/tsdb/chunks/head_chunks.go:439 +0x14f
github.com/prometheus/prometheus/tsdb.(*memSeries).mmapChunks(0xc00070e240, 0x18d83cd1a73?)
	/app/tsdb/head_append.go:1414 +0xba
github.com/prometheus/prometheus/tsdb.(*walSubsetProcessor).processWALSamples(0xc000aa0540, 0xc00071c400, 0xc0008d58b8?, 0x8b57ae?)
	/app/tsdb/head_wal.go:594 +0x49e
github.com/prometheus/prometheus/tsdb.(*Head).loadWAL.func2(0xc0008d5860?)
	/app/tsdb/head_wal.go:105 +0x45
created by github.com/prometheus/prometheus/tsdb.(*Head).loadWAL in goroutine 246
	/app/tsdb/head_wal.go:104 +0x307

```

### System information

Linux 6.1.72-96.166.amzn2023.x86_64 x86_64

### Prometheus version

```text
latest version of prometheus docker
```

### Prometheus configuration file

```yaml
global:
  scrape_interval:     10s

scrape_configs:
  - job_name: 'prod-1'
    static_configs:
      - targets: 
        - localhost:9100

  - job_name: 'prod-2'
    static_configs:
      - targets:
        - localhost:9101
              
  - job_name: "dev"
    static_configs:
      - targets: 
        - localhost:9099
```

### Logs

```text
goroutine 428 [running]:
github.com/prometheus/prometheus/tsdb.handleChunkWriteError({0x3f3f3a0?, 0xc0009f6cd8?})
	/app/tsdb/head_append.go:1432 +0x6e
github.com/prometheus/prometheus/tsdb/chunks.(*ChunkDiskMapper).WriteChunk(0xc000236000, 0xc000748d70?, 0x2760318?, 0xc001b11e90?, {0x3f73af8, 0xc0009fc400}, 0x0?, 0x39d0540)
	/app/tsdb/chunks/head_chunks.go:439 +0x14f
github.com/prometheus/prometheus/tsdb.(*memSeries).mmapChunks(0xc000176840, 0x18d83cd1a73?)
	/app/tsdb/head_append.go:1414 +0xba
github.com/prometheus/prometheus/tsdb.(*walSubsetProcessor).processWALSamples(0xc000d20540, 0xc000101800, 0xc0007b7af8?, 0x8b57ae?)
	/app/tsdb/head_wal.go:594 +0x49e
github.com/prometheus/prometheus/tsdb.(*Head).loadWAL.func2(0xc0007b7aa0?)
	/app/tsdb/head_wal.go:105 +0x45
created by github.com/prometheus/prometheus/tsdb.(*Head).loadWAL in goroutine 59
	/app/tsdb/head_wal.go:104 +0x307
ts=2024-02-07T14:00:27.450Z caller=main.go:544 level=info msg="No time or size retention was set so using the default time retention" duration=15d
ts=2024-02-07T14:00:27.450Z caller=main.go:588 level=info msg="Starting Prometheus Server" mode=server version="(version=2.49.1, branch=HEAD, revision=43e14844a33b65e2a396e3944272af8b3a494071)"
ts=2024-02-07T14:00:27.450Z caller=main.go:593 level=info build_context="(go=go1.21.6, platform=linux/amd64, user=root@6d5f4c649d25, date=20240115-16:58:43, tags=netgo,builtinassets,stringlabels)"
ts=2024-02-07T14:00:27.450Z caller=main.go:594 level=info host_details="(Linux 6.1.72-96.166.amzn2023.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 17 00:42:52 UTC 2024 x86_64 ip-172-31-6-235.eu-west-1.compute.internal (none))"
ts=2024-02-07T14:00:27.450Z caller=main.go:595 level=info fd_limits="(soft=65536, hard=65536)"
ts=2024-02-07T14:00:27.450Z caller=main.go:596 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2024-02-07T14:00:27.551Z caller=web.go:565 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
ts=2024-02-07T14:00:27.552Z caller=main.go:1039 level=info msg="Starting TSDB ..."
ts=2024-02-07T14:00:27.556Z caller=tls_config.go:274 level=info component=web msg="Listening on" address=[::]:9090
ts=2024-02-07T14:00:27.556Z caller=tls_config.go:277 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
ts=2024-02-07T14:00:27.848Z caller=dir_locker.go:77 level=warn component=tsdb msg="A lockfile from a previous execution already existed. It was replaced" file=/prometheus/data/lock
ts=2024-02-07T14:00:28.797Z caller=head.go:606 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2024-02-07T14:00:28.797Z caller=head.go:687 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=25.412µs
ts=2024-02-07T14:00:28.797Z caller=head.go:695 level=info component=tsdb msg="Replaying WAL, this may take a while"
panic: 2 errors: replace file: rename data/chunks_head/000031.tmp data/chunks_head/000031: operation not permitted; unlinkat data/chunks_head/000031.tmp: operation not permitted

goroutine 438 [running]:
github.com/prometheus/prometheus/tsdb.handleChunkWriteError({0x3f3f3a0?, 0xc000c82810?})
	/app/tsdb/head_append.go:1432 +0x6e
github.com/prometheus/prometheus/tsdb/chunks.(*ChunkDiskMapper).WriteChunk(0xc000790c30, 0xc0005d0d70?, 0x2760318?, 0xc001f9ec90?, {0x3f73af8, 0xc0009a96c0}, 0x30?, 0x39d0540)
	/app/tsdb/chunks/head_chunks.go:439 +0x14f
github.com/prometheus/prometheus/tsdb.(*memSeries).mmapChunks(0xc000748540, 0x18d83cd1a73?)
	/app/tsdb/head_append.go:1414 +0xba
github.com/prometheus/prometheus/tsdb.(*walSubsetProcessor).processWALSamples(0xc0006b7ef0, 0xc000612400, 0xc000c84a78?, 0x8b57ae?)
	/app/tsdb/head_wal.go:594 +0x49e
github.com/prometheus/prometheus/tsdb.(*Head).loadWAL.func2(0xc000c84a20?)
	/app/tsdb/head_wal.go:105 +0x45
created by github.com/prometheus/prometheus/tsdb.(*Head).loadWAL in goroutine 112
	/app/tsdb/head_wal.go:104 +0x307
```

**Comment by rapphil** [comment]
Hi,

I tried to reproduce your issue and I was not able to get exactly your same error, but I did noticed that there are some small details that you have to take care of when using fuse.

### TL;DR;

Are you mounting the bucket with the `allow_other` mount option?

To mount my s3 bucket, I used the following command: `s3fs <bucket> ~/volume/ -o allow_other`

Before running the mount command I had to edit the `/etc/fuse.conf` to add the following line:
```
user_allow_other
```
This will allow that other users, other than the one that mounted the mount point, to get access to the data, which seems to be a prereq in our case since the user running prometheus is different from the user creating the mount point.

After this I was able to make work properly and I can start/restart docker-compose without any errors.

### How to reproduce my setup

docker-compose.yaml
```
version: '3.9'

services:
  node-exporter:
    image: quay.io/prometheus/node-exporter:latest
    container_name: node_exporter
    command:
      - '--path.rootfs=/host'
    network_mode: host
    pid: host
    restart: unless-stopped
    volumes:
      - '/:/host:ro,rslave'

  prometheus:
    image: prom/prometheus:latest
    container_name: prometheus
    restart: unless-stopped
    network_mode: host
    volumes:
      - /home/ubuntu/volume/prometheus:/prometheus/
      - /home/ubuntu/volume/etc/prometheus:/etc/prometheus
    command:
      - '--config.file=/etc/prometheus/config.yaml'
    privileged: true
    user: "1000:1000"
```

etc/prometheus/config.yaml
```
global:
  scrape_interval:     10s

scrape_configs:
  - job_name: 'prod-1'
    static_configs:
      - targets:
        - localhost:9100
```
I installed s3fs using
```
sudo apt update
sudo apt install s3fs
```

To mount my s3 bucket, I used the following command: `s3fs <bucket> ~/volume/ -o allow_other`

Before running the mount command I had to edit the `/etc/fuse.conf` to add the following line:
```
user_allow_other
```

I'm using ubuntu 24.04 LTS on a EC2 host.


### Alternatives

It seems that you want to store long term data into a block storage to reduce costs. Have you considered using something like Cortex, Thanos or Mimir? .

* https://cortexmetrics.io/
* https://thanos.io/
* https://github.com/grafana/mimir

I think these are more reliable options since I don't know if there are any side effects of using s3fs as there are [limitations](https://github.com/s3fs-fuse/s3fs-fuse?tab=readme-ov-file#limitations).

**Comment by bboreham** [comment]
Hello, apologies for the delay.

I draw your attention to this note in [the documentation](https://prometheus.io/docs/prometheus/latest/storage/#operational-aspects):
> CAUTION: Non-POSIX compliant filesystems are not supported for Prometheus' local storage as unrecoverable corruptions may happen. 

From the error we can conclude that s3fs is not a POSIX-compliant filesystem.
